### PR TITLE
Enable prometheus metrics scraping

### PR DESCRIPTION
**Pull request by alebedev87**
Result:
image

### CODE REVIEW

**Comment by DhritiShikhar** [comment]
Should there be any tests for this @alebedev87 ?

File: Makefile [hunk 1]
```diff
@@ -156,6 +156,10 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
 	cd config/manager && $(KUSTOMIZE) edit set image quay.io/openshift/origin-external-dns-operator=${IMG}
 	# webhook volume and service are added explicilty so that they don't land in the bundle where it's managed by OLM
 	cd config/default && $(KUSTOMIZE) edit add patch --path=manager_webhook_volume_patch.yaml
+	# add service monitor with certificate from the service serving certificate
+	cd config/prometheus && $(KUSTOMIZE) edit add resource monitor.yaml
```
Comment (arjunrn) [inline Makefile:160]: Why not add the monitor resource directly or put another way include it in the default target.

File: README.md [hunk 2]
```diff
@@ -194,3 +195,20 @@ Prepare your environment for the installation commands.
 ## Proxy support
 
 [Configuring proxy support for ExternalDNS Operator](./docs/proxy.md)
+
+## Metrics
+
+The ExternalDNS Operator exposes [controller-runtime metrics](https://book-v1.book.kubebuilder.io/beyond_basics/controller_metrics.html) in the format expected by [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator).
```
Comment (arjunrn) [inline README.md:201]: Maybe link to the latest version of the [kubebuilder book](https://book.kubebuilder.io/reference/metrics.html)

File: README.md [hunk 3]
```diff
@@ -194,3 +195,20 @@ Prepare your environment for the installation commands.
...
+`ServiceMonitor` object is created in the operator's namespace (by default `external-dns-operator`), make sure that your instance of the Prometheus Operator is properly configured to find it.     
+You can check `.spec.serviceMonitorNamespaceSelector` and `.spec.serviceMonitorSelector` fields of `prometheus` resource and edit the operator's namespace or service monitor accordingly:
```
Comment (arjunrn) [inline README.md:203]: If the namespace can be add by adding a label then shouldn't we be adding the label to the namespace resource? Or is it not possible when the namespace is created by OLM.

File: config/default/kustomization.yaml [hunk 4]
```diff
@@ -6,16 +6,11 @@ resources:
 - ../rbac
 - ../manager
 - ../webhook
+- ../prometheus
```
Comment (arjunrn) [inline config/default/kustomization.yaml:9]: So since the _prometheus_ directory is already included do you need to `add` it with kustomize as well?

**Review by arjunrn — COMMENTED** [review]
Just a couple of questions. Otherwise looks good.

File: Makefile (hunk 1 above)
Comment (alebedev87) [reply Makefile]: Done, added `monitor.yaml` as default resource, `tlsConfig` patch is added for the bundle generation only.

File: config/default/kustomization.yaml (hunk 4 above)
Comment (alebedev87) [reply config/default/kustomization.yaml:9]: Not sure I understood this one. It's already in kustomize. You can deploy using `make deploy` or with a bundle. Both way will create `ServieMonitor` and all the necessary things.

**Comment by alebedev87** [comment]
@DhritiShikhar : for the e2e test we'd need a viable way of verifying the target creation in prometheus. Will try to find something.

**Comment by alebedev87** [comment]
As discussed with Dhriti, the e2e testing is possible but quite heavy. Prometheus targets are not exposed on the Kubernetes level (CRs or ConfigMaps/Secrets), they have to be retrieved from the prometheus API which most often is queried by `exec` inf into the prometheus POD.

File: README.md (hunk 3 above)
Comment (arjunrn) [reply README.md:204]: AFAIK the cluster prometheus should not be modified to monitor non cluster operator. Instead another prometheus instance should be used. The instructions are [here](https://docs.openshift.com/container-platform/4.10/monitoring/enabling-monitoring-for-user-defined-projects.html).

File: config/default/kustomization.yaml (hunk 4 above)
Comment (arjunrn) [reply config/default/kustomization.yaml:9]: My point is was that if the `ServiceMonitor` was already part of the kustomize do you need to `add` it again for the `make deploy` target.

**Pull request merged**